File: frontend/src/types/post.ts
import { User } from "./user";
import { Image } from "./image";

export interface PostItem {
  id: number;
  user: User;
  content: string;
  images?: Image[]; 
  timestamp: string; 
  likesCount: number; 
  commentsCount: number; 
  liked: boolean;
  saved: boolean;
  originalPost: Post | null;
  originalPostId: number | null;
  isDeleted?: boolean;
}

export interface Post {
  id: number;
  content: string;
  images: { 
    id: number;
    url: string;
  }[];
  createdAt: string;
  updatedAt: string | null;
  user: User;
  likesCount: number;
  liked: boolean;
  commentsCount: number;
  saved?: boolean;
  originalPost?: Post;
  deleted?: boolean;
}

File: backend/src/main/java/org/example/backend/entity/Post.java
package org.example.backend.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import jakarta.persistence.*;
import java.util.Date;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import org.hibernate.annotations.CreationTimestamp;

@Entity
@Table(name = "posts")
@Data
@NoArgsConstructor
@AllArgsConstructor 
@Builder
public class Post {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = true, columnDefinition = "TEXT")
    private String content;

    @JsonManagedReference // Thêm annotation này
    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Image> images;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false)
    @Temporal(TemporalType.TIMESTAMP)
    @CreationTimestamp
    private Date createdAt;

    @Column
    @Temporal(TemporalType.TIMESTAMP)
    private Date updatedAt;

    @ManyToOne
    @JoinColumn(name = "original_post_id")
    private Post originalPost; // Bài gốc nếu là repost

    @JsonManagedReference // Thêm annotation này
    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Comment> comments;

    @JsonManagedReference // Thêm annotation này
    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Like> likes;

    @Column(nullable = false)
    private boolean isDeleted = false;

    @Column
    @Temporal(TemporalType.TIMESTAMP)
    private Date deletedAt;

    @Transient
    private boolean likedByCurrentUser;

    @Transient
    private long likeCount;

    @Transient
    private long totalCommentCount;
}
